### ISSUE TITLE
Deadlock when cancelling ClientWebSocket.ConnectAsync

### ISSUE THREAD

**Issue by lkelley-bec** [dotnet/runtime#55013]
### Description

Sometimes cancelling the token passed into `ClientWebSocket.ConnectAsync()` causes a deadlock of some kind. The following NUnit test reproduces the problem in a random amount of time (on my machine, it's typically anywhere from 5-50 times when I run it in a bash loop):
```C#
[Test]
public async Task TestCancelWebSocketConnection()
{
    var cts = new CancellationTokenSource();
    var uri = new Uri("ws://localhost");
    var ws = new ClientWebSocket();
    var connectTask = ws.ConnectAsync(uri, cts.Token);
    cts.Cancel();
    try
    {
        await connectTask;
        Assert.AreNotEqual(WebSocketState.Open, ws.State);
    }
    catch (TaskCanceledException) { }
    catch (WebSocketException) { }
}
```

With this test, the call to `cts.Cancel()` will occasionally never resolve, and the call to `ws.ConnectAsync()` will get stuck.

The expected behavior is for the connection attempt to be cancelled if already in progress instead of deadlocking.

### Configuration

* .NET version: I've tried this targeting both .NET core 3.1 and .NET 5.0, with similar results (`dotnet --version` reports 5.0.301)
* OS: Ubuntu 20.04.2 LTS 64-bit
* Architecture: Running in VMWare with x86_64

### Regression?

Unknown, though it's present in both 3.1 and 5.0 as mentioned above.

### Other information

This sounds eerily similar to #16810, though I can't speak to whether the implementation concerns that issue describes apply here.

A slightly less minimal though seemingly more consistent snippet to reproduce is the following:
```C#
[Test]
[Timeout(5000)]
public void TestCancelWebSocketConnection()
{
    var cts = new CancellationTokenSource();
    var uri = new Uri("ws://localhost");
    async Task LoopConnect()
    {
        for (int i = 0; i < 3; i++)
        {
            try
            {
                var ws = new ClientWebSocket();
                await ws.ConnectAsync(uri, cts.Token);
            }
            catch (TaskCanceledException)
            {
                throw;
            }
            catch (Exception e)
            {
                if (cts.Token.IsCancellationRequested)
                {
                    throw new TaskCanceledException();
                }
                // loop
            }
        }
    }
    var connectTask = LoopConnect();
    cts.Cancel();
    Assert.ThrowsAsync<TaskCanceledException>(() => connectTask);
}
```

**Comment by lkelley-bec** [comment]
For others finding this issue, a potential workaround would be to artificially delay the cancellation by a bit of time (building off the initial example in the issue description):
```C#
    var cts = new CancellationTokenSource();
    var uri = new Uri("ws://localhost");
    var ws = new ClientWebSocket();

    // workaround: delay the cancellation
    var delayedCts = new CancellationTokenSource();
    cts.Token.Register(() => delayedCts.CancelAfter(100)); // not sure on the time here, but 100 ms seems to prevent the issue for me

    var connectTask = ws.ConnectAsync(uri, delayedCts.Token);
    cts.Cancel();
```
This does seem to indicate that the problem is some kind of race condition that happens if `ConnectAsync` is _immediately_ cancelled.

**Comment by CarnaViire** [comment]
@lkelley-bec unfortunately I was unable to reproduce that neither on Windows nor on Linux... Could you please share the dump of the process when deadlock happens?

**Comment by lkelley-bec** [comment]
Couldn't attach the dump file directly (maybe it was too big?), so I made a separate repo and uploaded it there:
https://github.com/lkelley-bec/dotnet-issue-55013-repro

**Comment by stephentoub** [comment]
Is this the right process?  It seems like this is a dump of the test runner process, which has spawned another process and is now just waiting for it:
![image](https://user-images.githubusercontent.com/2642209/126576200-e5ca9a87-54e6-4b07-80a8-1b886521521c.png)

**Comment by lkelley-bec** [comment]
Apologies, I didn't realize there were multiple processes running. I still can't attach files here, so I pushed a [new commit](https://github.com/lkelley-bec/dotnet-issue-55013-repro/commit/4e5d49b139dd793363c3b2d0aa5e8a889d532f6f) with a replacement dump that looks to actually include the ClientWebSocket and my test code.

**Comment by stephentoub** [comment]
Thanks for sharing the new dump.

It's definitely a deadlock.  The socket connect code was problematically invoking a callback while holding a lock, something that's rarely a good idea:
https://github.com/dotnet/runtime/blob/release/5.0/src/libraries/System.Net.Sockets/src/System/Net/Sockets/MultipleConnectAsync.cs#L90-L131
with DnsCallback invoking SocketAsyncEventArgs.OnCompleted which in turn ran the await continuation.  This was in response to the web socket connect request failing ("Connection refused"), and the finally block in the socket connect code then tries to Dispose the CancellationTokenRegistration:
https://github.com/dotnet/runtime/blob/74905e767eb7b971636310965b66bf5dfd6a3545/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs#L109-L112
If when you call CancellationTokenRegistration.Dispose, the associated cancellation callback is currently in flight, Dispose will wait for it to complete, so as to satisfy the guarantee that all associated work with that registration will not run after Dispose returns.
![image](https://user-images.githubusercontent.com/2642209/126654454-a01e9062-0d68-472c-8caa-9972308f0bcb.png)'

And, indeed, cancellation is currently in-flight, invoking Socket.CancelConnectAsync, which in turn invokes MultipleConnectAsync.Cancel, which tries to take the exact same lock that's held by DoDnsCallback in the other thread:
https://github.com/dotnet/runtime/blob/74905e767eb7b971636310965b66bf5dfd6a3545/src/libraries/System.Net.Sockets/src/System/Net/Sockets/MultipleConnectAsync.cs#L316-L320
![image](https://user-images.githubusercontent.com/2642209/126654779-6b83d6a7-f4c1-44cf-a732-4ae20816e632.png)

And that's the deadlock: Thread 9857 is holding the lock while blocking waiting for the cancellation callback running on Thread 9866 to complete, and the cancellation callback on thread 9866 is blocked waiting for the lock held by Thread 9857.

The good news is that https://github.com/dotnet/runtime/pull/43661 rewrote all of this code, entirely deleting MultiConnectAsync. Can you try to repro on .NET 6?  I expect it won't repro.

**Comment by lkelley-bec** [comment]
Given that it typically fails before 50 loops on .NET 5 and it's just passed loop 1000 on 6.0.0-preview.6 without getting stuck, it seems to be fixed there. Thanks for the quick find!

What's the best path forward for users of .NET 3.1? I can probably get away with either the workaround I mentioned earlier (which really only mitigates the problem, if I understand your explanation correctly) or just not use a `CancellationToken` at all if need be.

**Comment by karelz** [comment]
@lkelley-bec we would consider servicing 3.1 if there is strong need for that -- i.e. no workarounds, high impact on more customers, etc.
Is that the case for you?
Given that 6.0 is around the corner, would upgrading to 6.0 be an option for you as well?

Closing as addressed in 6.0

**Comment by lkelley-bec** [comment]
Waiting for 6.0 will work for my situation. Thank you for your help!
